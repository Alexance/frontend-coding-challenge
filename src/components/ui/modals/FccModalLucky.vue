<script setup lang="ts">
import { ElButton, ElDialog } from "element-plus";

const emit = defineEmits<{
  (event: "update:modelValue", newValue: boolean): void;
  (event: "reject" | "accept"): void;
}>();
</script>

<template>
  <ElDialog :model-value="true" @close="emit('reject')" width="30%" center>
    <span>
      Congratulations, `<slot name="name"></slot>` is a lucky name! You've won a
      prize. Will you accept it?
    </span>

    <template #footer>
      <div class="dialog-footer">
        <ElButton @click="emit('reject')">Reject</ElButton>
        <ElButton type="primary" @click="emit('accept')">Accept</ElButton>
      </div>
    </template>
  </ElDialog>
</template>
