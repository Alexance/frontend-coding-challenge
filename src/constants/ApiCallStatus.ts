export enum ApiCallStatus {
  NOT_AVAILABLE = "NOT_AVAILABLE",
  SUCCESS = "SUCCESS",
  FAILURE = "FAILURE",
}
