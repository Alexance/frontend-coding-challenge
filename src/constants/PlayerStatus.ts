export enum PlayerStatus {
  NO_DETAILS = "NO_DETAILS",
  NO_WIN = "NO_WIN",
  WIN = "WINNER",
}
