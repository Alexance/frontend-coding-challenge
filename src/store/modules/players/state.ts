import { Person } from "@/models/Person";
import { PlayerStatus as PlayerStatus } from "@/constants/PlayerStatus";
import { ApiCallStatus } from "@/constants/ApiCallStatus";

export interface State {
  /**
   * Current user information
   */
  user: {
    name: string;
  };
  /**
   * The list of all winners
   */
  allWinners: Person[];
  /**
   * The list of today's winners
   */
  todayWinners: Person[];
  /**
   * If the given user is a winner or has another player status
   */
  status: PlayerStatus;
  /**
   * The status of the request to get all winners
   */
  getWinnersRequestStatus: ApiCallStatus;
}
