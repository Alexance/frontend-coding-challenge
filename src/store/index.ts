import { InjectionKey } from "vue";
import { createStore, Store, useStore as useVuexStore } from "vuex";

import playersModule from "./modules/players";
import { RootState } from "./state";

export const key: InjectionKey<Store<RootState>> = Symbol();

export const useStore = () => {
  return useVuexStore(key);
};

export default createStore({
  modules: {
    players: playersModule,
  },
});
