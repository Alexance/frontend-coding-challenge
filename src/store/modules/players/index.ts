import { Module } from "vuex";

import { RootState } from "@/store/state";
import { PlayerStatus } from "@/constants/PlayerStatus";
import { ApiCallStatus } from "@/constants/ApiCallStatus";

import { GetterTypes } from "./getter-types";
import { MutationTypes } from "./mutation-types";
import { ActionTypes } from "./action-types";
import { State } from "./state";
import getters from "./getters";
import mutations from "./mutations";
import actions from "./actions";

const storeModule: Module<State, RootState> = {
  state: (): State => ({
    user: {
      name: "",
    },
    todayWinners: [],
    allWinners: [],
    status: PlayerStatus.NO_DETAILS,
    getWinnersRequestStatus: ApiCallStatus.NOT_AVAILABLE,
  }),

  getters,
  mutations,
  actions,
};

export { State, GetterTypes, MutationTypes, ActionTypes };
export default storeModule;
