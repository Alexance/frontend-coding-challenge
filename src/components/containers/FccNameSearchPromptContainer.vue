<script setup lang="ts">
import { computed } from "vue";

import { Person } from "@/models/Person";
import { PlayerStatus } from "@/constants/PlayerStatus";
import { useStore } from "@/store";
import { MutationTypes } from "@/store/modules/players";
import FccModalLucky from "@/components/ui/modals/FccModalLucky.vue";
import FccModalUnlucky from "@/components/ui/modals/FccModalUnlucky.vue";
import FccModalDuplicate from "@/components/ui/modals/FccModalDuplicate.vue";
import FccModalLimitExceeded from "../ui/modals/FccModalLimitExceeded.vue";

const store = useStore();

const status = computed(() => store.state.players.status);
const name = computed(() => store.state.players.user.name);

function onAccept() {
  // Direct state access (instead of a getter) is reasonable here
  // because the full user object is necessary only once, it's not used
  // anywhere and the array of all winners is wiped out each time the
  // form is submitted
  const user = store.state.players.allWinners.find(
    (winner) => winner.name === name.value
  ) as Person;

  store.commit(MutationTypes.ADD_TODAY_WINNER, user);
  store.commit(MutationTypes.SET_PLAYER_STATUS, PlayerStatus.NO_DETAILS); // To hide the modal window
  store.commit(MutationTypes.SET_USER_NAME, "");
}

function onReject() {
  store.commit(MutationTypes.SET_PLAYER_STATUS, PlayerStatus.NO_DETAILS);
}
</script>

<template>
  <FccModalLucky
    v-if="status === PlayerStatus.WIN"
    @accept="onAccept"
    @reject="onReject"
    data-test="fcc-modal-lucky"
  >
    <template #name>{{ name }}</template>
  </FccModalLucky>

  <FccModalUnlucky
    v-else-if="status === PlayerStatus.NO_WIN"
    @close="onReject"
    data-test="fcc-modal-unlucky"
  >
    <template #name>{{ name }}</template>
  </FccModalUnlucky>

  <FccModalDuplicate
    v-else-if="status === PlayerStatus.DUPLICATE"
    @close="onReject"
  >
    <template #name>{{ name }}</template>
  </FccModalDuplicate>

  <FccModalLimitExceeded
    v-else-if="status === PlayerStatus.LIMIT_EXCEEDED"
    @close="onReject"
  />
</template>
