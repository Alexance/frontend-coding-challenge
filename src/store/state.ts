import { State as PlayersState } from "./modules/players";

export interface RootState {
  players: PlayersState;
}
