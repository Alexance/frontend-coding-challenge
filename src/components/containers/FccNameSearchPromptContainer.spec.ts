import { reactive } from "vue";
import { Store } from "vuex";
import { shallowMount } from "@vue/test-utils";

import { key } from "@/store";
import { RootState } from "@/store/state";
import { ApiCallStatus } from "@/constants/ApiCallStatus";
import { PlayerStatus } from "@/constants/PlayerStatus";
import FccModalLucky from "@/components/ui/modals/FccModalLucky.vue";

import FccNameSearchPromptContainer from "./FccNameSearchPromptContainer.vue";
import { MutationTypes } from "@/store/modules/players";

describe("Given the `FccNameSearchPromptContainer` component", () => {
  // This mock should be defined as a common one and moved to
  // a separate test utility function or component
  const storeMock: Store<RootState> = {
    commit: jest.fn(),
    dispatch: jest.fn(),
    getters: {},
    hasModule: jest.fn().mockReturnValue(true),
    hotUpdate: jest.fn(),
    install: jest.fn(),
    registerModule: jest.fn(),
    replaceState: jest.fn(),
    state: reactive({
      players: {
        allWinners: [
          {
            name: "Ava",
          },
        ],
        getWinnersRequestStatus: ApiCallStatus.NOT_AVAILABLE,
        status: PlayerStatus.NO_DETAILS,
        todayWinners: [],
        yesterdayWinners: [],
        user: {
          name: "Ava",
        },
      },
    }),
    subscribe: jest.fn(),
    subscribeAction: jest.fn(),
    watch: jest.fn(),
    unregisterModule: jest.fn(),
  };
  const render = () =>
    shallowMount(FccNameSearchPromptContainer, {
      global: {
        provide: {
          // There is type mismatch issue in the library.
          // It doesn't allow to pass the Injection Key (Symbol)
          [key as any]: storeMock,
        },
      },
    });
  let wrapper: ReturnType<typeof render>;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe("when it is rendered", () => {
    beforeEach(() => {
      wrapper = render();
    });

    describe("when the user is a winner", () => {
      beforeAll(() => {
        storeMock.state.players.status = PlayerStatus.WIN;
      });

      describe("and when a user rejects the prompt", () => {
        beforeEach(() => {
          wrapper
            .findComponent<typeof FccModalLucky>(
              '[data-test="fcc-modal-lucky"]'
            )
            // This direct component instance access fixes @vue/test-utils
            // issue when it returns the parent component by the selector,
            // not a correct child
            .getComponent(FccModalLucky)
            .vm.$emit("reject");
        });

        it("should close the window", () => {
          expect(storeMock.commit).toHaveBeenCalledWith(
            MutationTypes.SET_PLAYER_STATUS,
            PlayerStatus.NO_DETAILS
          );
        });

        it("should keep the name is in field", () => {
          expect(storeMock.commit).not.toHaveBeenCalledWith(
            MutationTypes.SET_USER_NAME,
            expect.any(String)
          );
        });
      });

      describe("and when a user accepts the prompt", () => {
        beforeEach(() => {
          wrapper
            .findComponent<typeof FccModalLucky>(
              '[data-test="fcc-modal-lucky"]'
            )
            // This direct component instance access fixes @vue/test-utils
            // issue when it returns the parent component by the selector,
            // not a correct child
            .getComponent(FccModalLucky)
            .vm.$emit("accept");
        });

        it("should close the window", () => {
          expect(storeMock.commit).toHaveBeenCalledWith(
            MutationTypes.SET_PLAYER_STATUS,
            PlayerStatus.NO_DETAILS
          );
        });

        it("should wipe the name field value", () => {
          expect(storeMock.commit).toHaveBeenCalledWith(
            MutationTypes.SET_USER_NAME,
            expect.any(String)
          );
        });

        it("should add the player to the list of the winners", () => {
          expect(storeMock.commit).toHaveBeenCalledWith(
            MutationTypes.ADD_TODAY_WINNER,
            storeMock.state.players.allWinners[0]
          );
        });
      });
    });

    describe("when the user is NOT a winner", () => {
      beforeAll(() => {
        storeMock.state.players.status = PlayerStatus.NO_WIN;
      });

      it("should match the snapshot", () => {
        expect(wrapper.element).toMatchSnapshot();
      });
    });
  });
});
