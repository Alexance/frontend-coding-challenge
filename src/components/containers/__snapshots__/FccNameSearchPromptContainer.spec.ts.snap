// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the \`FccNameSearchPromptContainer\` component when it is rendered when the user is NOT a winner should match the snapshot 1`] = `
<fcc-modal-unlucky-stub
  data-test="fcc-modal-unlucky"
/>
`;
